Deploying microservice apps with MongoDB and Elasticsearch to Heroku doesn't work

##### **Overview of the issue**

Error on startup is:

```
2018-09-20T11:47:34.316188+00:00 app[web.1]: Caused by: java.lang.IllegalStateException: duplicate key: data
```

You can see the problem if you look at `application-heroku.yml`:

```yaml
spring:
    data:
        mongodb:
            uri: ${MONGODB_URI}?authMode=scram-sha1
    data:
        jest:
            uri: ${BONSAI_URL}
```

##### **Motivation for or Use Case**

Deploying a microservice to Heroku should work.

##### **Reproduce the error**

Follow the steps [described in this gist](https://gist.github.com/mraible/3c97f60c6617e67fd87cd9895d54311c), then try to deploy the "store" app to Heroku.

##### **Suggest a Fix**

Change from:

```yaml
spring:
    data:
        mongodb:
            uri: ${MONGODB_URI}?authMode=scram-sha1
    data:
        jest:
            uri: ${BONSAI_URL}
```

To:

```
spring:
    data:
        mongodb:
            uri: ${MONGODB_URI}?authMode=scram-sha1
        jest:
            uri: ${BONSAI_URL}
```

##### **JHipster Version(s)**

5.3.4

##### **JHipster Version(s)**

```
store@0.0.0 /Users/mraible/dev/okta/okta-jhipster-microservices-oauth-example/store
└── generator-jhipster@5.3.4

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "applicationType": "microservice",
    "clientPackageManager": "npm",
    "databaseType": "mongodb",
    "devDatabaseType": "mongodb",
    "enableHibernateCache": false,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "jhiPrefix": "jhi",
    "languages": [
      "en",
      "fr"
    ],
    "messageBroker": false,
    "nativeLanguage": "en",
    "packageName": "com.okta.developer.store",
    "packageFolder": "com/okta/developer/store",
    "prodDatabaseType": "mongodb",
    "searchEngine": "elasticsearch",
    "serviceDiscoveryType": "eureka",
    "skipClient": true,
    "skipServer": false,
    "testFrameworks": [],
    "useSass": false,
    "websocket": false,
    "baseName": "store",
    "authenticationType": "oauth2",
    "serverPort": "8082",
    "jhipsterVersion": "5.3.4",
    "buildTool": "maven",
    "cacheProvider": "hazelcast",
    "clientFramework": "angularX",
    "skipUserManagement": true,
    "herokuAppName": "okta-store",
    "herokuDeployType": "git"
  },
  "entities": [
    "Product"
  ]
}
</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

The passed validation must be valid to be added to the field.
Errors: No value
events.js:183
      throw er; // Unhandled 'error' event
      ^

Error: ERROR!
Error while parsing entities to JDL

    at Environment.error (/Users/mraible/dev/okta/okta-jhipster-microservices-oauth-example/store/node_modules/yeoman-environment/lib/environment.js:157:40)
    at module.exports.error (/Users/mraible/dev/okta/okta-jhipster-microservices-oauth-example/store/node_modules/generator-jhipster/generators/generator-base.js:2060:18)
    at module.exports.generateJDLFromEntities (/Users/mraible/dev/okta/okta-jhipster-microservices-oauth-example/store/node_modules/generator-jhipster/generators/generator-base-private.js:1000:18)
    at module.exports.displayEntities (/Users/mraible/dev/okta/okta-jhipster-microservices-oauth-example/store/node_modules/generator-jhipster/generators/info/index.js:57:34)
    at Object.<anonymous> (/Users/mraible/dev/okta/okta-jhipster-microservices-oauth-example/store/node_modules/yeoman-generator/lib/index.js:418:27)
    at /Users/mraible/dev/okta/okta-jhipster-microservices-oauth-example/store/node_modules/run-async/index.js:25:25
    at new Promise (<anonymous>)
    at /Users/mraible/dev/okta/okta-jhipster-microservices-oauth-example/store/node_modules/run-async/index.js:24:19
    at self.env.runLoop.add.completed (/Users/mraible/dev/okta/okta-jhipster-microservices-oauth-example/store/node_modules/yeoman-generator/lib/index.js:419:13)
    at runCallback (timers.js:810:20)

--- a/generators/heroku/templates/application-heroku.yml.ejs
+++ b/generators/heroku/templates/application-heroku.yml.ejs
@@ -49,12 +49,8 @@ spring:
         hikari:
             maximumPoolSize: 8
 <%_ } else if (prodDatabaseType === 'mongodb') { _%>
-    data:
-        mongodb:
-            uri: ${MONGODB_URI}?authMode=scram-sha1
+    data.mongodb.uri: ${MONGODB_URI}?authMode=scram-sha1
 <%_ } _%>
 <%_ if (searchEngine === 'elasticsearch') { _%>
-    data:
-        jest:
-            uri: ${BONSAI_URL}
+    data.jest.uri: ${BONSAI_URL}
 <%_ } _%>

--- a/generators/heroku/templates/application-heroku.yml.ejs
+++ b/generators/heroku/templates/application-heroku.yml.ejs
@@ -48,9 +48,15 @@ spring:
         password: ${JDBC_DATABASE_PASSWORD}
         hikari:
             maximumPoolSize: 8
-<%_ } else if (prodDatabaseType === 'mongodb') { _%>
-    data.mongodb.uri: ${MONGODB_URI}?authMode=scram-sha1
 <%_ } _%>
-<%_ if (searchEngine === 'elasticsearch') { _%>
-    data.jest.uri: ${BONSAI_URL}
+<%_ if (prodDatabaseType === 'mongodb' || searchEngine === 'elasticsearch') { _%>
+    data:
+  <%_ if (prodDatabaseType === 'mongodb') { _%>
+        mongodb:
+            uri: ${MONGODB_URI}?authMode=scram-sha1
+  <%_ } _%>
+  <%_ if (searchEngine === 'elasticsearch') { _%>
+        jest:
+            uri: ${BONSAI_URL}
+  <%_ } _%>
 <%_ } _%>

--- a/test/heroku.spec.js
+++ b/test/heroku.spec.js
@@ -257,6 +257,7 @@ describe('JHipster Heroku Sub Generator', () => {
                 assert.file(expectedFiles.monolith);
                 assert.fileContent('.yo-rc.json', '"herokuDeployType": "jar"');
                 assert.fileContent(`${constants.SERVER_MAIN_RES_DIR}/config/application-heroku.yml`, 'datasource:');
+                assert.fileContent(`${constants.SERVER_MAIN_RES_DIR}/config/application-heroku.yml`, 'jest');
                 assert.noFileContent(`${constants.SERVER_MAIN_RES_DIR}/config/application-heroku.yml`, 'mongodb:');
             });
         });

--- a/test/heroku.spec.js
+++ b/test/heroku.spec.js
@@ -257,7 +257,8 @@ describe('JHipster Heroku Sub Generator', () => {
                 assert.file(expectedFiles.monolith);
                 assert.fileContent('.yo-rc.json', '"herokuDeployType": "jar"');
                 assert.fileContent(`${constants.SERVER_MAIN_RES_DIR}/config/application-heroku.yml`, 'datasource:');
-                assert.fileContent(`${constants.SERVER_MAIN_RES_DIR}/config/application-heroku.yml`, 'jest');
+                assert.fileContent(`${constants.SERVER_MAIN_RES_DIR}/config/application-heroku.yml`, 'jest:');
+                assert.fileContent(`${constants.SERVER_MAIN_RES_DIR}/config/application-heroku.yml`, 'data:');
                 assert.noFileContent(`${constants.SERVER_MAIN_RES_DIR}/config/application-heroku.yml`, 'mongodb:');
             });
         });

